### PR TITLE
Allow the message frame to create chunks as large as transport will allow

### PR DESCRIPTION
**Pull request by louiscryan**
For https://github.com/grpc/grpc-java/issues/312

### CODE REVIEW

**Comment by louiscryan** [comment]
FYI  - This PR is work in progress, initially just demonstrates idea.

**Comment by nmittler** [comment]
@louiscryan One potential issue I see with this is that  `OutputStreamAdapter` which calls `writeRaw` for every call to `write(byte[],int, int)`, which could lead to a lot of buffer allocations. I guess this won't be a problem in most cases since we'll be using one of the `Deferred` input streams, which I think would make a single call to `write`.

It would be interesting to see if the number of allocations increases significantly with this change.

**Comment by ejona86** [comment]
Note that right now this guarantees two buffer allocations per message: 1 for the header and 1 for the body. That isn't all that hard to fix though, so it can still be useful in seeing how things behave for really large messages.

**Comment by louiscryan** [comment]
@nmittler The buffer allocator can and probably should choose to allocate a buffer with a minimum size in the case where the messages are small to address this issue when we fallback to ByteStreams.copy for the degenerate case.

Similarly the allocator can place an upper bound on the size of the buffer allocated so we don't fragment buffer pools or thrash the heap for very large messages.

The other optimization for more sophisticated bindings is to allow the frame to receive messages that are already in the right buffer format for the transport layer. Right now we only admit byte[] in flushTo but we could let a deferred stream act as a factory for transport buffers

**Comment by louiscryan** [comment]
@ejona86 yes, using one alloc for the header and the message instead of two does make a difference too. For context some numbers from running QpsClient with 16k payloads, 10s warmup, 20s run

baseline --> 8574 qps
one alloc for header & body --> 9082 qps
one alloc for header & body & alloc message len --> 10429 qps

which is a pretty decent improvement

**Comment by louiscryan** [comment]
@ejona86 Ready for review

File: integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 package io.grpc.testing.integration;
 
 import io.grpc.ChannelImpl;
+import io.grpc.transport.netty.NegotiationType;
```
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java]: Revert the changes to this file?

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java [hunk 2]
```diff
@@ -38,8 +38,24 @@
  * The default allocator for {@link OkHttpWritableBuffer}s used by the OkHttp transport.
  */
 class OkHttpWritableBufferAllocator implements WritableBufferAllocator {
+
+  private final int maxDataLength;
+
+  /**
+   * Construct a new instance
+   * @param maxDataLength the maximum allowed length of a DATA frame.
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java]: Instead of limiting here, could we do min(maxBytes, maxDataLength) in [OutboundFlowController.OutboundFlowState.writeBytes()](https://github.com/grpc/grpc-java/blob/af188767134e387a798803230c730d9f99ed214f/okhttp/src/main/java/io/grpc/transport/okhttp/OutboundFlowController.java#L316)

Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java]: Hmm... maybe it is better to do this here, as the frame.split() may choose to copy (if we get misaligned between Buffer's Segments and the HTTP/2 streams).

**Comment by ejona86** [comment]
@louiscryan, LGTM, after reverting changes to Http2NettyTest. After reverting that Travis should turn green.

File: core/src/main/java/io/grpc/transport/MessageFramer.java [hunk 3]
```diff
@@ -181,7 +184,8 @@ private void writeRaw(byte[] b, int off, int len) {
         commitToSink(false);
       }
       if (buffer == null) {
-        buffer = bufferAllocator.allocate(maxFrameSize);
+        // Allocate as big a buffer as the allocator will allow
```
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: Is this comment correct?  Shouldn't it be something like "Allocate a buffer that is at least big enough to for this write"?

File: core/src/main/java/io/grpc/transport/WritableBufferAllocator.java [hunk 4]
```diff
@@ -32,13 +32,14 @@
...
+ * An allocator of buffers provided by the transport implementation to {@link MessageFramer} so
+ * it can send chunks of data to the transport in a form that the transport can directly serialize.
  */
 public interface WritableBufferAllocator {
 
   /**
-   * Allocate a new {@link WritableBuffer} that can hold
-   * {@code capacity} bytes.
+   * Request a new {@link WritableBuffer} with the given {@code capacity}. The allocator is
+   * free to return a buffer with a greater or lesser capacity.
```
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: It can return a lesser capacity?

Comment (ejona86) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: We are wanting to use the size as a hint. If that hint is 10 MB, we may want to give back 256K or 1M instead. If the hint is 1 byte, we may want to give back 2K or 4K. MessageFramer doesn't have any requirements on the returned buffer size, other than it being able to hold at least one byte.

Comment (nmittler) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: That's not yet being done by this PR, correct?  Is the idea that `NettyWritableBufferAllocator` would be modified to be aware of the max frame size and therefore possibly provide a buffer smaller than what was requested?

Once flow control kicks in, however, all bets are off since whatever we pass to the transport will likely get chunked anyway.  From the perspective of the framer writing to the transport, bigger is always better ... which means we probably don't need to care too much about the current max frame size WRT the framer. Maybe we just always request a buffer that's something a bit larger (since they're pooled), like 1MB?

Comment (ejona86) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: Yes, it isn't done in this PR. Yes, `NettyWritableBufferAllocator` would be modified to be aware of the maximum _pooled_ buffer size.

This doesn't actually have anything to do with HTTP/2's frame sizes (for Netty; for OkHttp we are using it for that purpose). Instead, once the allocation is too large Netty will no longer pool the Buffer, which likely outweighs the cost of larger buffers.

My initial problem with this PR was that it seemed we would always make sure to use a single buffer able to hold the entire message. Being able to "max out" the requested size at 256K or 1M is the primary reason I'm okay with this PR now.

Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: So there are a number of step-functions in the pooled buffer allocator that make this choice interesting and I suspect we'll do a couple of rounds on this and we probably want benchmarks to dictate whether it's better to to send larger buffers into the flow-controller or smaller.

I'm running Netty's benchmarks to inform the decision for the lower-bound but the choice for the upper-bound is more interesting...

- Picking an upper-bound that is guaranteed to fit within the arena does not guarantee that we get a pooled buffer (the arena might be exhausted)
- Taking buffers out of the arena as opposed to the buffer pool (> 32k) means larger locks so there is a tradeoff between the cost of those locks and the cost of main-heap allocation.
- If we do enforce an upper bound we will consume aggressively from the pool. A single very large object could exhaust the pool and starve smaller allocations. This is the opposite of what we want as the pool is designed to optimize smaller allocs.

I'm inclined for the moment to set an upper-bound but in truth we need benchmarking to determine (a) what it should be and (b) if we really need it at all.

Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: OK results of benchmark run are here

https://gist.github.com/louiscryan/c5997da99c51b5893a00

note that 'default' indicates the use of thread-local buffer caching. Some observations

- Pooled >>> Unpooled, particularly for larger buffers.
- Thread-local caching is masking the cost of arena locks which grow as buffer sizes grow, once we cross the 32k buffer cache boundary we see a moderate step in performance.
- After the 32k boundary allocation costs are relatively constant irrespective of buffer size until we exceed the area limit of 16MB. At which point performance absolutely stinks.

It seems like a toss-up between 4k & 16k  for the minimum size. For the maximum size we can either stop at 32k to limit to cacheable-buffers or we can go up to the arena size max. Given that the step-function around 32k is only 2x it seems reasonable to go higher, say 1MB to being with.

File: integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java (hunk 1 above)
Comment (louiscryan) [inline integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java]: Done

File: core/src/main/java/io/grpc/transport/WritableBufferAllocator.java (hunk 4 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: I'd favor 4k over 16k for the minimum size. We expect many applications to use messages < 4k exclusively, so choosing 16k would just cause waste. Stopping at 1 MB sounds fine.

Why is there such a high variation between the two sets of tests:
ByteBufAllocatorBenchmark.defaultPooledDirectAllocAndFree   65536  thrpt   20   2409135.087 ± 136903.037  ops/s
ByteBufAllocatorBenchmark.defaultPooledDirectAllocAndFree   0065536  thrpt   20  3024966.732 ±  83142.423  ops/s

File: core/src/main/java/io/grpc/transport/MessageFramer.java (hunk 3 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: Not exactly no, see discussion below. Renamed param in interface to 'capacityHint'

**Comment by louiscryan** [comment]
PTAL

File: integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java (hunk 1 above)
Comment (ejona86) [inline integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java]: Unused import is the only change to this file.

Comment (louiscryan) [inline integration-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java]: Done

File: core/src/main/java/io/grpc/transport/WritableBufferAllocator.java (hunk 4 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: Ill flip it back to 4k.

The variation is caused by having to reduce the number of outstanding buffers when increasing the allocated capacity to such a large value to prevent OOMs

File: netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java [hunk 5]
```diff
@@ -33,12 +33,28 @@
...
 
 /**
- * The default allocator for {@link NettyWritableBuffer}s used by the Netty transport.
+ * The default allocator for {@link NettyWritableBuffer}s used by the Netty transport. We set a
+ * minimum bound to avoid unnecessary re-allocation for small follow-on writes and to facilitate
+ * Netty's caching of buffer objects for small writes. We set an upper-bound to avoid allocations
+ * outside of the arena-pool which are orders of magnitude slower. The Netty transport can receive
+ * buffers of arbitrary size and will chunk them based on flow-control so there is no transport
+ * requirement for an upper bound.
+ * <p>
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java]: This should be on the same line as note:
\<p\>Note:

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java (hunk 2 above)
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java]: Looking at this some more, it doesn't seem there will be any performance difference between the two. So I do think I would prefer us improving OutboundFlowState.writeBytes().

File: core/src/main/java/io/grpc/transport/WritableBufferAllocator.java (hunk 4 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/WritableBufferAllocator.java]: @louiscryan, but the 64K test that was part of the large capacity test was _faster._

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java (hunk 2 above)
Comment (louiscryan) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java]: OK. Will move into OutboundFlowState

Comment (louiscryan) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java]: This largely means it should handle buffers > maxframesize. We still need an upper bound here however which can still be 4k for the moment.

File: netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java (hunk 5 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java]: Done

File: okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java (hunk 2 above)
Comment (louiscryan) [inline okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpWritableBufferAllocator.java]: Done.

**Comment by louiscryan** [comment]
PTAL

**Comment by ejona86** [comment]
@louiscryan, LGTM, after fixing Travis failure (checkstyle failure).

**Comment by louiscryan** [comment]
Merged. Note that the performance numbers shown above depend on the ability to send larger messages in a single DATA frame and thus need larger flow-control windows. The ability to configure that is coming in a separate PR from @buchgr
